File: lua/ht/plugins/external/rest-nvim.lua
return {
  Use {
    "rest-nvim/rest.nvim",
    lazy = {
      opts = {
        skip_ssl_verification = true,
      },
      dependencies = {
        "nvim-lua/plenary.nvim",
      },
      config = function(_, opts)
        require("rest-nvim").setup(opts)

        local right_click_items = {
          {
            "Http",
            children = {
              {
                "Run the request under cursor",
                callback = function()
                  require("rest-nvim").run()
                end,
                keys = "r",
              },
              {
                "Preview the request cURL command",
                callback = function()
                  require("rest-nvim").run(true)
                end,
                keys = "p",
              },
              {
                "Rerun the last command",
                callback = function()
                  require("rest-nvim").last()
                end,
                keys = "l",
              },
            },
            keys = "t",
          },
        }

        local RC = require("ht.core.right-click")
        RC.add_section {
          index = RC.indexes.http,
          enabled = {
            filetype = {
              "http",
            },
          },
          items = right_click_items,
        }
      end,
    },
    functions = {
      {
        filter = {
          filter = function(buffer)
            if buffer.filename == nil then
              return false
            end
            return buffer.filename:match(".*%.http") ~= nil
          end,
        },
        values = {
          FuncSpec("Run the request under cursor", function()
            require("rest-nvim").run()
          end),
          FuncSpec("Preview the request cURL command", function()
            require("rest-nvim").run(true)
          end),
          FuncSpec("Rerun the last command", function()
            require("rest-nvim").last()
          end),
        },
      },
    },
  },
}

File: src/conf/plugins/other/rest-nvim.ts
import { RightClickIndexes } from "@conf/base";
import {
  ActionGroupBuilder,
  Plugin,
  PluginOptsBase,
  andActions,
} from "@core/model";

function generateActions() {
  return ActionGroupBuilder.start()
    .from("rest.nvim")
    .category("Rest")
    .condition((buf) => {
      let [ret] = string.match(buf.fullFileName, ".*%.http");
      return ret !== null;
    })
    .addOpts({
      id: "rest-nvim.run-request",
      title: "Run the request under cursor",
      callback: () => {
        luaRequire("rest-nvim").run();
      },
    })
    .addOpts({
      id: "rest-nvim.preview-curl-command",
      title: "Preview the request cURL command",
      callback: () => {
        luaRequire("rest-nvim").run(true);
      },
    })
    .addOpts({
      id: "rest-nvim.rerun-last-command",
      title: "Re-run the last command",
      callback: () => {
        luaRequire("rest-nvim").last();
      },
    })
    .build();
}

const spec: PluginOptsBase = {
  shortUrl: "rest-nvim/rest.nvim",
  lazy: {
    opts: {
      skip_ssl_verification: true,
    },
    ft: ["http"],
    lazy: true,
    dependencies: ["nvim-lua/plenary.nvim"],
    config: (_, opts) => {
      luaRequire("rest-nvim").setup(opts);
    },
  },
  allowInVscode: true,
};

export const plugin = new Plugin(andActions(spec, generateActions));
